pkgs/development/interpreters/python/hooks/python-relax-build-deps-hook.py

```diff
@@ -0,0 +1,42 @@
+from os import environ
+from packaging.requirements import Requirement
+from packaging.specifiers import SpecifierSet
+from pathlib import Path
+import json, tomlkit
+
+
+nixAttrs = json.load(Path(environ['NIX_ATTRS_JSON_FILE']).open())
+removeDeps = frozenset(nixAttrs.get("pythonRemoveBuildDeps", []))
+
+relaxDeps = nixAttrs.get("pythonRelaxBuildDeps", {})
+if isinstance(relaxDeps, list):
+    relaxDeps = { req.name: req for req in map(Requirement, relaxDeps) }
+else:
+    def requirementWithSpec(name, spec):
+        """Construct a Requirement with a given name and version specifier"""
+        r = Requirement(name)
+        r.specifier = SpecifierSet(spec)
+        return r
+
+    relaxDeps = {
+        name: requirementWithSpec(name, spec)
+        for name, spec in relexDeps.items()
+    }
```

Comment:
While I like the idea of going `pythonRelaxBuildDeps.foo = ">1, <2";` or such, especially if version constraints need to be (in part) programatically generated

... I'm not sure the usecase is there at the moment, and it comes with non-negligible complexity (~20% of the code, and among the less nice-to-read parts)

Reply:
I take that back: the list form permits writing something like `[ "foo =1" "foo =2" ]`, which would be messier to even just find and error-out on.

While the attrset form diverges from `pythonRelaxDeps`' (current) API, it entirely obviate this issue, so I believe this should be the only accepted form instead.

FWIW, it looks like `pythonRelaxDepsHook` does not check either for redundant entries.

What do you think, @ShamrockLee ?

Reply:
> While I like the idea of going `pythonRelaxBuildDeps.foo = ">1, <2";` or such, especially if version constraints need to be (in part) programatically generated

Sounds great!

For such interface, are you planning to handle `pythonRelaxDeps.foo = null` as a way to "cancel" `foo`'s relaxation?

Reply:
@mweinelt What do you think about the attribute set interface?

Reply:
From experience, I mostly just remove the constraint instead of mutating it. We just need to deal with overzealous pinning here.